### DISCUSSION TITLE
foreign spelling in uppercase

### DISCUSSION THREAD

**Discussion by eadmaster** [polm/cutlet#38] [Q&A] [answered]
is it possible to get foreign spellings in uppercase?

````
import cutlet
katsu = cutlet.Cutlet()
katsu.use_foreign_spelling_uppercase = True
katsu.romaji("カツカレーは美味しい")
# => 'CUTLET CURRY wa oishii'
````

**Comment by polm** [comment]
It's not currently possible to do that. I don't think it would make sense as a hard-coded feature, but it would be easy to add information to output tokens about whether they use foreign spelling or not, for example.

Out of curiosity, what situation would you use this in?

**Comment by eadmaster** [comment]
> Out of curiosity, what situation would you use this in?

with this [lrctransliterate script](https://github.com/eadmaster/dotfiles/blob/b5125e3e46950f82fb870c7e4829c9b20837f0e9/bin/lrctransliterate.py), to better differentiate between native and foreign words.

**Reply by polm** [reply]
OK, I added a feature for this in #39, you can try that out and see if it fulfills your needs before I merge it.

**Reply by eadmaster** [reply]
ok, thanks, can you post me an example usage?

**Reply by polm** [reply, marked as answer]
```
import cutlet

katsu = cutlet.Cutlet()

text = "カツカレーを食べた"

tokens = katsu.romaji_tokens(katsu.tagger(text))

def upper_foreign(tok):
    if tok.foreign:
        return tok.surface.upper()
    else:
        return tok.surface

res = [upper_foreign(tok) for tok in tokens]
print(*res)
```

**Reply by eadmaster** [reply]
thanks, it works fine.

is it also possible to detect words that are in foreign spelling already in the input text?
e.g.
````
text = "First kissから始まる" 
# => 'FIRST KISS kara hajimaru'
````

**Reply by polm** [reply]
You can use `char_type` in the MeCab token features to tell if a token is alphabetic (5).

```
import cutlet
katsu = cutlet.Cutlet()

for tok in katsu.tagger("First kissから始まる"):
  print(tok.char_type, tok.surface)
```

**Reply by eadmaster** [reply]
is there a way to obtain the original untransliterated token from the same loop?

````python
	tokens = trans_obj.romaji_tokens(trans_obj.tagger(line_text), capitalize=False)
	for tok in tokens:	
		if tok.foreign:  # --> or tok.original_surface.isascii() 
			s += tok.surface.upper()
		else:
			s += tok.surface
````

EDIT: nevermind, i've found a way using 2 separate loops over the tokens.

**Comment by eadmaster** [comment]
i did some extra testing with my lyrics collection and found a line that is throwing an exception:
````
Traceback (most recent call last):
  File "/home/andy/bin/lrctransliterate.py", line 109, in <module>
    line_text = jap_trans(trans_obj, line_text)
  File "/home/andy/bin/lrctransliterate.py", line 48, in jap_trans
    tokens = trans_obj.romaji_tokens(trans_obj.tagger(line_text), capitalize=False)
  File "/home/andy/.local/lib/python3.10/site-packages/cutlet/cutlet.py", line 224, in romaji_tokens
    roma = self.romaji_word(word)
  File "/home/andy/.local/lib/python3.10/site-packages/cutlet/cutlet.py", line 325, in romaji_word
    return self.map_kana(kana)
  File "/home/andy/.local/lib/python3.10/site-packages/cutlet/cutlet.py", line 370, in map_kana
    out += self.get_single_mapping(pk, char, nk)
  File "/home/andy/.local/lib/python3.10/site-packages/cutlet/cutlet.py", line 423, in get_single_mapping
    return self.table[kk]
KeyError: 'ﾞ'
````

the line was
`(What!)誰かの声かﾞ`

(the [updated script is here](https://github.com/eadmaster/dotfiles/blob/main/bin/lrctransliterate.py#L30))

**Reply by polm** [reply]
You need to clean the text by calling `cutlet.normalize_text(text)` before passing it to the tagger to avoid edge cases like this. That's mentioned in the docstring, I omitted it from the example for brevity.
